### ISSUE TITLE
[KDA] five coverage blind spots let the t_inv defect (#54813) pass every existing test

### ISSUE THREAD

**Issue by kgrujcic-tt** [tenstorrent/tt-metal#54839]
Sub-issue of #54814. Hit during Kimi-K3 bring-up (#53508).

#54813 was a total failure — the KDA layer uncorrelated with its own reference on 68 of Kimi-K3's 69 KDA layers — and every existing test passed. Five independent blind spots, each worth fixing on its own terms because the same shapes will hide the next defect.

**1. `t_inv` is scored by whole-tensor PCC.** It is `I + strictly-lower`, so PCC is dominated by the identity diagonal. The broken output scored **0.99508**, above the suite's own 0.999 bar, while its strictly-lower part — the entire content of the UT transform — scored **0.935**. Any threshold on this output belongs on the strictly-lower part, not the whole tile.

**2. `_host_inputs` pins the gate to `[-0.051, -0.001]`.** Kimi-K3's gate is `KDA_GATE_LOWER_BOUND * sigmoid(...)` with a lower bound of -5.0, spanning `(-5, 0)` — two orders of magnitude outside coverage, and exactly the regime the defect lives in.

**3. `host_protocol` pins every other magnitude too:** `final_decay` to [0.86, 0.94], the strict-lower block feeding `t_inv` to 0.015, `kd`/`k_dec_t` to 0.025. Every scan case draws from one point in a seven-dimensional magnitude space.

**4. `make_kimi_k3_test_case` pairs real weights with `torch.randn` activations.** `test_kimi_k3_layer_1_real_weights_pcc` exercises the exact layer the defect breaks and passes at 0.9995. **Nothing tested real weights against real activations** — the only combination where blind spots 2/3 and the real data meet.

**5. The `invert_doubling` comment stated its own stability claim "must be validated empirically."** It never was, so the defect sat documented-as-unverified in-tree for the life of the op.

Two op-level suites covering the magnitudes (1)-(3) never generate were written during the bring-up and handed over in the comments below. They are not in-tree, and they need re-validating against `invert_block_ps4` (#54937), which #55420 shows is itself incomplete.

**Comment by kgrujcic-tt** [comment]
Handing over the two suites the body's last line refers to. They are dropped from #54850 (Kimi-K3 bring-up, ds-prefill area) because they are KDA op coverage, not K3 work, and they belong with the op.

Both were written against `invert_horner`, which #54937 has since replaced with `invert_block_ps4` (exact, 8 matmuls, +1.38%). **They need re-validating against that inverse before landing** - specifically the `t_inv` row in the first file, whose numbers in the docstring are pre-#54937.

| file | covers | blind spot |
|---|---|---|
| `test_prepare_chunk_recurrence_gate_range.py` | `prepare_chunk_recurrence` on Kimi-K3's real gate, all 7 outputs scored separately | 1, 2, 4 |
| `test_recurrent_chunk_scan_magnitudes.py` | one protocol scale swept at a time against the suite's own oracle | 3 |

Neither needs a model, a checkpoint or a mesh: both drive the op directly and score against the suite's existing oracle. Drop them under `tests/ttnn/nightly/unit_tests/operations/experimental/kda/`.

One caveat carried in the first file's docstring and worth repeating here: widening the synthetic gate to K3's full span makes the **oracle** overflow first (at span 5.0 its `t_inv` is entirely NaN), so any PCC there measures the reference breaking, not the kernel. That is why it drives the op with real layer-1 tensors instead, the one layer where the oracle stays finite and the end-to-end result is still wrong.

<details><summary><code>test_prepare_chunk_recurrence_gate_range.py</code></summary>

```python
# SPDX-FileCopyrightText: © 2026 Tenstorrent USA, Inc.
# SPDX-License-Identifier: Apache-2.0
"""`prepare_chunk_recurrence` on Kimi-K3's real gate, against the suite's own oracle.

Every existing case builds the gate as `g = -0.001 - 0.05 * rand`, so the op is only ever asked for
`g` in [-0.051, -0.001]. Kimi-K3's gate is `KDA_GATE_LOWER_BOUND * sigmoid(...)` with a lower bound
of -5.0, and its real per-32-row-chunk cumulative sum reaches -56 at layer 1 and -160 at layer 0.
That is far outside the covered range, and it is where the chunk formulation gets delicate: `intra`
and `t_inv` both pair `exp(cumsum g)` against `exp(-cumsum g)`, whose product is bounded by 1 for
i >= j but whose factors separately span `exp(±|cumsum|)`.

Note what this file does NOT claim. A synthetic sweep widening the gate to K3's full span makes the
ORACLE overflow first — at span 5.0 its `t_inv` is entirely NaN and `intra` carries 159 infinities —
so any PCC there measures the reference breaking, not the kernel. `exp(160)` is not representable in
fp32 at all, so the naive chunk formulation simply has no valid answer at layer 0's range, and only
a formulation that forms `exp(cumsum_i - cumsum_j)` directly does.

So this drives the op with the model's real tensors at the one layer where the oracle stays finite,
which is exactly the layer that fails end-to-end: layer 1 scores ~0 against its own torch reference
on device while layer 0 scores 0.99993. Each of the seven outputs is scored separately, so a single
wrong one names the stage.

Read the `t_inv` row carefully. It scored 0.99508 before the `invert_horner` fix, which looks like a
pass and is not one: `t_inv` is `I + strictly-lower`, so whole-tensor PCC is dominated by the
identity diagonal and stays near 1 while the off-diagonals — the entire content of the UT transform
— are wrong. Scored on its strictly-lower part alone it was 0.935, and substituting it for the torch
protocol took the recurrence from 1.0 to 0.0014. Any future threshold on this output should be set
on the strictly-lower part, not the whole tile.
"""

from __future__ import annotations

from pathlib import Path

import pytest
import torch
import torch.nn.functional as F
from loguru import logger

import ttnn
from models.common.utility_functions import comp_pcc, run_for_blackhole
from models.demos.deepseek_v3_d_p.reference.kda.ops import causal_depthwise_conv_reference, kda_gate_reference
from models.demos.deepseek_v3_d_p.reference.kimi_k3_config import kimi_k3_kda_config
from models.demos.deepseek_v3_d_p.tests.kda.checkpoint_utils import load_kda_layer_state_dict
from models.demos.deepseek_v3_d_p.tests.kimi_k3.golden import TRACE_100K, resolve_checkpoint, resolve_trace
from tests.ttnn.nightly.unit_tests.operations.experimental.kda.test_prepare_chunk_recurrence import (
    CHUNK_SIZE,
    OUTPUT_NAMES,
    _device_inputs,
    _oracle,
    _run,
)

pytestmark = [run_for_blackhole()]

SEQ = 1024
BF16_MASK = 0x26


def _real_inputs(checkpoint: Path, hidden: torch.Tensor, layer_idx: int, config):
    """q, k, v, g, beta for one real Kimi-K3 layer, in the flat layout the op takes."""
    weights = load_kda_layer_state_dict(checkpoint, layer_idx, config)
    hidden = hidden.float().unsqueeze(0)
    state = hidden.new_zeros(1, config.conv_kernel_size - 1, config.q_dim)

    def projected(name, conv):
        out, _ = causal_depthwise_conv_reference(
            F.linear(hidden, weights[f"{name}.weight"].float()), weights[conv], state
        )
        return out

    q, k, v = (projected(n, f"{n[0]}_conv1d.weight") for n in ("q_proj", "k_proj", "v_proj"))
    raw = F.linear(F.linear(hidden, weights["f_a_proj.weight"].float()), weights["f_b_proj.weight"].float())
    gate = kda_gate_reference(
        raw.reshape(1, hidden.shape[1], config.num_heads, config.head_k_dim),
        weights["A_log"],
        weights["dt_bias"],
        config.gate_lower_bound,
    ).reshape(1, hidden.shape[1], config.q_dim)
    beta = torch.sigmoid(F.linear(hidden, weights["b_proj.weight"].float()))
    beta = beta.reshape(hidden.shape[1] // CHUNK_SIZE, CHUNK_SIZE, config.num_heads).permute(2, 0, 1).unsqueeze(-1)
    return q, k, v, gate, beta.contiguous()


@run_for_blackhole()
def test_prepare_matches_oracle_on_real_kimi_k3_gate(device):
    checkpoint = resolve_checkpoint()
    trace = resolve_trace(TRACE_100K)
    if checkpoint is None or trace is None:
        pytest.skip("needs KIMI_K3_HF_MODEL and the 100k golden trace")
    config = kimi_k3_kda_config()
    hidden = trace.rows("kda", "kda_input_layer_0", 0, SEQ)

    for layer_idx in (0, 1):
        inputs = _real_inputs(Path(checkpoint), hidden, layer_idx, config)
        cumsum = inputs[3].reshape(1, SEQ // CHUNK_SIZE, CHUNK_SIZE, -1).cumsum(dim=2)
        try:
            expected = _oracle(inputs, config.num_heads, BF16_MASK)
        except Exception as error:
            # Layer 0's chunk cumsum reaches -160, so exp(-cumsum) is inf and the Gram matrix the
            # oracle inverts is singular. The reference formulation has no answer here at all, which
            # is a fact about the formulation and not about the kernel: the device reproduces this
            # layer at 0.99993 against the stable recurrent reference.
            logger.info(
                f"  layer {layer_idx} (chunk cumsum min {float(cumsum.min()):7.1f}): "
                f"oracle unrepresentable -- {type(error).__name__}: {error}"
            )
            continue
        finite = all(torch.isfinite(t.float()).all() for t in expected)
        actual = _run(_device_inputs(inputs, device), config.num_heads, output_bf16_mask=BF16_MASK)
        scores = "  ".join(
            f"{name} {float(str(comp_pcc(want.float(), ttnn.to_torch(got).float(), 0.99)[1]).split()[-1]):8.5f}"
            for name, want, got in zip(OUTPUT_NAMES, expected, actual, strict=True)
        )
        logger.info(f"  layer {layer_idx} (chunk cumsum min {float(cumsum.min()):7.1f}, oracle finite={finite})")
        logger.info(f"    {scores}")
```

</details>

<details><summary><code>test_recurrent_chunk_scan_magnitudes.py</code></summary>

```python
# SPDX-FileCopyrightText: © 2026 Tenstorrent USA, Inc.
# SPDX-License-Identifier: Apache-2.0
"""Does `recurrent_chunk_scan` hold up outside the magnitudes its suite generates?

`host_protocol` pins every tensor to one scale — `final_decay` to [0.86, 0.94], the strict-lower
block feeding `t_inv` to 0.015, `kd`/`k_dec_t` to 0.025 — and every existing test draws from it. So
the op is covered at exactly one point in a seven-dimensional magnitude space.

That matters because Kimi-K3's real weights land somewhere else in that space, and there the KDA
layer disagrees with its own torch reference completely: layer 0 scores 0.99993 against the model's
recorded output while all 68 other KDA layers score ~0. The failing variable is the decay magnitude
and nothing else — shifting layer 0's `dt_bias` down by 1.0 takes it from 0.99994 to -0.005, and
shifting layer 1's up by 3.0 takes it from 0.004 to 0.99905. Dtype is not the cause (forcing every
preparation output to fp32 changes nothing), nor is math fidelity, chunk grouping, or sequence
length: it is already wrong with a single chunk per rank.

This sweeps one protocol scale at a time against the suite's own oracle, so a failure here is a
reproduction with no model, no checkpoint and no mesh in it.
"""

from __future__ import annotations

import pytest
import torch
from loguru import logger

import ttnn
from models.common.utility_functions import comp_pcc, run_for_blackhole
from tests.ttnn.nightly.unit_tests.operations.experimental.kda.recurrent_chunk_scan_test_utils import (
    CHUNK_SIZE,
    device_protocol,
    host_protocol,
    initial_state,
    recurrent_oracle,
    run_recurrent,
    to_device,
)

pytestmark = [run_for_blackhole()]

BATCH_HEADS, NUM_CHUNKS, KEY_DIM, VALUE_DIM = 8, 4, 128, 128


def _scaled(protocol, index, factor):
    out = list(protocol)
    out[index] = out[index] * factor
    return tuple(out)


def _with_final_decay(protocol, base):
    """Replace the [0.86, 0.94] band with one centred on `base`, keeping the same spread shape."""
    out = list(protocol)
    generator = torch.Generator().manual_seed(99)
    out[5] = (base + 0.08 * torch.rand(*out[5].shape, generator=generator)).float()
    return tuple(out)


@run_for_blackhole()
def test_scan_matches_oracle_across_magnitudes(device):
    state = initial_state(BATCH_HEADS, KEY_DIM, VALUE_DIM)
    base_protocol = host_protocol(BATCH_HEADS, NUM_CHUNKS, KEY_DIM, VALUE_DIM)

    def check(label, protocol):
        want_out, want_state = recurrent_oracle(protocol, state)
        got = run_recurrent(device_protocol(protocol, device), to_device(state, device))
        pcc_out = float(str(comp_pcc(want_out.float(), ttnn.to_torch(got[0]).float(), 0.99)[1]).split()[-1])
        pcc_state = float(str(comp_pcc(want_state, ttnn.to_torch(got[1]).float(), 0.99)[1]).split()[-1])
        logger.info(f"  {label:38s} output {pcc_out:9.5f}   state {pcc_state:9.5f}")

    check("suite default", base_protocol)
    for base in (0.86, 0.6, 0.4, 0.2, 0.05, 0.98):
        check(f"final_decay base {base:.2f}", _with_final_decay(base_protocol, base))
    for name, index in (("v_beta", 0), ("kd", 1), ("q_decay", 2), ("intra", 3), ("k_dec_t", 4), ("t_inv", 6)):
        for factor in (4.0, 16.0):
            check(f"{name} x{factor:g}", _scaled(base_protocol, index, factor))
```

</details>

**Comment by kgrujcic-tt** [comment]
The re-validation flagged above is now #55420: `invert_block_ps4` degrades 10 of 18 Kimi-K3 KDA layers on a single chip (worst 0.7508), and takes the model to 0.0009 from layer 5 on.

Both suites here cover exactly the magnitudes that break it, which makes them the regression gate for whatever fixes #55420 rather than just coverage debt.
